src/signature.ts

```diff
@@ -1,31 +1,98 @@
-const { ecdsaSign, ecdsaRecover, publicKeyConvert } = require('ethereum-cryptography/secp256k1')
+import { ecdsaSign, ecdsaRecover, publicKeyConvert } from 'ethereum-cryptography/secp256k1'
 import BN from 'bn.js'
 import { toBuffer, setLengthLeft, bufferToHex, bufferToInt } from './bytes'
 import { keccak } from './hash'
 import { assertIsBuffer } from './helpers'
-import { BNLike } from './types'
+import { BNLike, PrefixedHexString } from './types'
+import { isHexString } from '.'
 
 export interface ECDSASignature {
   v: number
   r: Buffer
   s: Buffer
 }
 
+export interface ECDSASignatureBN {
+  v: BN
+  r: Buffer
+  s: Buffer
+}
+
+export interface ECDSASignatureBuffer {
+  v: Buffer
+  r: Buffer
+  s: Buffer
+}
+
+export interface ECDSASignatureHexString {
+  v: PrefixedHexString
+  r: Buffer
+  s: Buffer
+}
+
 /**
  * Returns the ECDSA signature of a message hash.
  */
-export const ecsign = function(
+export function ecsign(msgHash: Buffer, privateKey: Buffer, chainId?: number): ECDSASignature
+export function ecsign(msgHash: Buffer, privateKey: Buffer, chainId: BN): ECDSASignatureBN
+export function ecsign(msgHash: Buffer, privateKey: Buffer, chainId: Buffer): ECDSASignatureBuffer
+export function ecsign(
   msgHash: Buffer,
   privateKey: Buffer,
-  chainId?: number
-): ECDSASignature {
+  chainId: PrefixedHexString
+): ECDSASignatureHexString
+export function ecsign(msgHash: Buffer, privateKey: Buffer, chainId: any): any {
   const sig = ecdsaSign(msgHash, privateKey)
   const recovery: number = sig.recid
 
-  const ret = {
-    r: Buffer.from(sig.signature.slice(0, 32)),
-    s: Buffer.from(sig.signature.slice(32, 64)),
-    v: chainId ? recovery + (chainId * 2 + 35) : recovery + 27
+  let ret
```

Comment:
Can we explicitly add types here to `ret` so we know in which `if` block we are using which return value? So ret would be of type `ECDSASignature | ECDSASignatureBN | ECDSASignatureBuffer` (possibly have to add `undefined` to keep TypeScript happy).

Reply:
refactored the code here, should be good now, please feel free to comment if it could still be improved